eks/multicam_smoother.py

```diff
@@ -11,39 +14,182 @@
+
+
+def fit_eks_mirrored_multicam(
+    input_source: Union[str, list],
+    save_file: str,
+    bodypart_list: Optional[list] = None,
+    smooth_param: Optional[Union[float, list]] = None,
+    s_frames: Optional[list] = None,
+    camera_names: Optional[list] = None,
+    quantile_keep_pca: float = 95,
+    avg_mode: str = 'median',
+    zscore_threshold: float = 2,
+) -> tuple:
+    """
+    Fit the Ensemble Kalman Smoother for mirrored multi-camera data.
+
+    Args:
+        input_source: Directory path or list of CSV file paths with columns for all cameras.
+        save_file: File to save output DataFrame.
+        bodypart_list: List of body parts.
+        smooth_param: Value in (0, Inf); smaller values lead to more smoothing.
+        s_frames: Frames for automatic optimization if smooth_param is not provided.
+        camera_names: List of camera names corresponding to the input data.
+        quantile_keep_pca: Percentage of points kept for PCA (default: 95).
+        avg_mode: Mode for averaging across ensemble ('median', 'mean').
+        zscore_threshold: Z-score threshold for filtering low ensemble std.
+        save_dir: The directory path to save output files to.
+
+    Returns:
+        tuple: Smoothed DataFrames for each camera, final smoothing parameters, and NLL values.
+    """
+    # Load and format input files
+    input_dfs_list, output_df, keypoint_names = format_data(input_source)
+    if bodypart_list is None:
+        bodypart_list = keypoint_names
+
+    # loop over keypoints; apply eks to each individually
+    for keypoint_ensemble in bodypart_list:
+        # Separate body part predictions by camera view
+        marker_list_by_cam = [[] for _ in range(len(camera_names))]
+        for markers_curr in input_dfs_list:
+            for c, camera_name in enumerate(camera_names):
+                non_likelihood_keys = [
+                    key for key in markers_curr.keys()
+                    if camera_names[c] in key and keypoint_ensemble in key
+                ]
+                marker_list_by_cam[c].append(markers_curr[non_likelihood_keys])
+
+        # Run the ensemble Kalman smoother for multi-camera data
+        camera_dfs, smooth_params_final, nll_values = ensemble_kalman_smoother_multicam(
+            markers_list_cameras=marker_list_by_cam,
+            keypoint_ensemble=keypoint_ensemble,
+            smooth_param=smooth_param,
+            quantile_keep_pca=quantile_keep_pca,
+            camera_names=camera_names,
+            s_frames=s_frames,
+            ensembling_mode=avg_mode,
+            zscore_threshold=zscore_threshold,
+        )
+        # Put results in new dataframe
+        for camera_name in camera_names:
+            output_df = populate_output_dataframe(
+                camera_dfs[camera_name],
+                keypoint_ensemble,
+                output_df,
+                key_suffix=f'_{camera_name}'
+            )
+        # Save the output DataFrames to CSV file
```

Comment:
should de-indent these two lines so `output_df.to_csv(save_file)` is only called once